#########################################################################
# ⭐⭐⭐  Set parameters carefully for satisfied performance!   ⭐⭐⭐ #
#########################################################################

mass: 1.4065 # kg
gra: 9.81
pose_solver: 1 # 0:From ZhepeiWang (drag & less singular) 1:From ZhepeiWang, 2:From rotor-drag
ctrl_freq_max: 100.0
use_bodyrate_ctrl: true
max_manual_vel: 1.0
max_angle: 80 # Attitude angle limit in degree. A negative value means no limit.
low_voltage: 13.2 # 4S battery
odom_freq: 100
ctrl_mode: 2 # DFBC 1 MPC 2 JPCM 3

odom_topic: /Odom_high_freq
imu_topic: /mavros/imu/data
target_topic: /planning/pos_cmd
attitude_target_topic: /mavros/setpoint_raw/attitude

rc_reverse: # *
    roll: false
    pitch: true
    yaw: true
    throttle: true

auto_takeoff_land:
    enable: true
    enable_auto_arm: true
    no_RC: true # If true, the auto takeoff/land will be executed even if RC is connected.
    takeoff_height: 1.5 # m
    takeoff_land_speed: 0.20 # m/s

thrust_model: # The model that maps thrust signal thr_bodyrate_u(0~1) to real thrust force F(Unit:N): F=K1*Voltage^K2*(K3*thr_bodyrate_u^2+(1-K3)*thr_bodyrate_u).
    print_value: false # display the value of “thr_scale_compensate” or “hover_percentage” during thrust model estimating.
    accurate_thrust_model: false # This can always enabled if don't require accurate control performance :-)
    # accurate thrust mapping parameters
    K1: 0.7583 # Needs precise calibration!
    K2: 1.6942 # Needs precise calibration!
    K3: 0.6786 # Needs precise calibration! K3 equals THR_MDL_FAC in https://docs.px4.io/master/en/config_mc/pid_tuning_guide_multicopter.html.
    # approximate thrust mapping parameters
    hover_percentage: 0.3365 # Thrust percentage in Stabilize/Arco mode # *
    thrust_upper_bound: 0.50
    thrust_lower_bound: 0.00

gain:
    # Thrust PID
    Kp0: 1.5
    Kp1: 1.5
    Kp2: 2.0
    Kv0: 1.0
    Kv1: 1.0
    Kv2: 1.5
    # Kp0: 4.0
    # Kp1: 4.0
    # Kp2: 5.0
    # Kv0: 2.0
    # Kv1: 2.0
    # Kv2: 3.0
    # ↓↓↓ No use now --
    Kvi0: 0.0
    Kvi1: 0.0
    Kvi2: 0.0
    Kvd0: 0.0
    Kvd1: 0.0
    Kvd2: 0.0
    # ↓↓↓ Only used in rate control mode.
    KAngR: 5.0
    KAngP: 5.0
    KAngY: 4.0
    # max error
    PErrMax: 1.0
    VErrMax: 1.0

rotor_drag:
    # 阻力系数向量
    # x: 0.24  # The reduced acceleration on each axis caused by rotor drag. Unit:(m*s^-2)/(m*s^-1).
    # y: 0.22  # Same as above
    # z: 0.31  # Same as above
    # k_thrust_horz: 0.005 #0.007 # Set to 0 recommended... --
    x: 0. # The reduced acceleration on each axis caused by rotor drag. Unit:(m*s^-2)/(m*s^-1).
    y: 0. # Same as above
    z: 0. # Same as above
    k_thrust_horz: 0.00 #0.007 # Set to 0 recommended... --

msg_timeout:
    odom: 0.5
    rc: 0.5
    cmd: 0.1
    imu: 0.5
    bat: 0.5

Factor_graph:
    LOG_NAME: TGYRO

    PRI_VICON_POS_COV: 0.001 # 位置测量的协方差（来自 Vicon 等传感器的测量噪声）
    PRI_VICON_VEL_COV: 0.001 # 速度测量的协方差（来自 Vicon 等传感器的测量噪声）

    # 位置偏差的协方差
    CONTROL_P_COV_X: 0.15
    CONTROL_P_COV_Y: 0.15
    CONTROL_P_COV_Z: 0.2
    # 最后一个时刻的位置偏差的协方差
    CONTROL_PF_COV_X: 0.03
    CONTROL_PF_COV_Y: 0.03
    CONTROL_PF_COV_Z: 0.03
    # 姿态偏差的协方差（3 维，三个轴的姿态约束强度
    CONTROL_R3_COV: 0.03
    CONTROL_R2_COV: 0.4
    CONTROL_R1_COV: 0.4
    # 期望轨迹速度的协方差
    CONTROL_V_COV: 1.3

    # 动力学模型的噪声模型
    DYNAMIC_P_COV: 0.0001
    DYNAMIC_R_COV: 0.0001
    DYNAMIC_V_COV: 0.0001
    # 输入加加速度（Jerk）的噪声模型
    INPUT_JERK_T: 0.2
    INPUT_JERK_M: 0.1
    INPUT_JERK_M3: 0.01
    # 控制输入（推力和陀螺力矩）的物理极限，用于自定义因子
    CLF_HIGH: 20 # 推力上限
    CLF_LOW: 5 # 推力下限
    G_CLF_HIGH: 3 # 陀螺力矩上限
    G_CLF_LOW: -3 # 陀螺力矩下限
    CLF_THR: 1 # 阈值
    G_CLF_THR: 0.1 # 阈值
    CLF_ALPHA: 1 # 0~1, 越大越平滑

    OPT_LENS_TRAJ: 20
    WINDOW_SIZE: 5

    POS_MEAS_MEAN: 0.000 # gps测量噪声均值
    POS_MEAS_COV: 0.500 # gps测量噪声协方差
    VEL_MEAS_COV: 0.050 # 测量噪声协方差
    ROT_MEAS_COV: 0.001 # 测量噪声协方差

    # imu预积分噪声协方差
    PRIOR_POS_MEAS_COV: 0.01
    PRIOR_VEL_MEAS_COV: 0.50
    PRIOR_ROT_MEAS_COV: 0.10

    # IMU Params
    acc_sigma_x: 0.1 # 0.3
    acc_bias_imu_x: 0.00004

    acc_sigma_y: 0.1
    acc_bias_imu_y: 0.001

    acc_sigma_z: 0.1
    acc_bias_imu_z: 0.001

    gyro_sigma_x: 0.01 # 0.02
    gyro_bias_sigma_x: 0.000018

    gyro_sigma_y: 0.03
    gyro_bias_sigma_y: 0.001

    gyro_sigma_z: 0.01
    gyro_bias_sigma_z: 0.0001

    # acc_sigma:        0.01
    # gyro_sigma:       0.001
    # acc_bias_imu:     0.001
    # gyro_bias_sigma:  0.0001

    prior_acc_sigma: 0.01
    prior_gyro_sigma: 0.01

    opt_gravity_rot: false
    prior_gravity_roll: 0.00
    prior_gravity_pitch: 0.00

    use_vel: false
    use_rot: true
